db/transactions.go

```diff
@@ -21,75 +19,96 @@ func NewTxManager(db *sqlx.DB) *TxManager {
 	return &TxManager{DB: db}
 }
 
-func (s *TxManager) WithTransaction(ctx context.Context, f func(context.Context, Querier) error) (err error) {
-	ctx, span := o11y.StartSpan(ctx, "tx-manager: with-transaction")
-	defer o11y.End(span, &err)
+type queryFn func(context.Context, Querier) error
 
-	// Retry this transaction 3 times
-	for i := 0; i < 3; i++ {
-		err = s.WithOneTransaction(ctx, f)
-		if !errors.Is(err, ErrBadConn) {
-			break
-		}
-		o11y.AddField(ctx, "bad_con", i)
-		o11y.AddField(ctx, "warning", err)
-	}
+// WithTransaction wraps f in an explicit o11y'd transaction, handling rollback
+// if f returns an error. It will retry the transaction a few times in the face of
+// ErrBadConn errors.
+// The length here is due to the internalised func, which we want to encapsulate
+// to avoid reuse, since it is highly coupled to the retry behaviour.
+//nolint:funlen
+func (s *TxManager) WithTransaction(ctx context.Context, f queryFn) (err error) {
+	const maxAttempts = 3
 
-	// Note that the above defer can reassign err
-	return err
-}
+	// Set up the main transaction function that we will retry on ErrBadCon
+	transaction := func(attempt int) (err error) {
+		ctx, span := o11y.StartSpan(ctx, "tx-manager: tx")
+		defer o11y.End(span, &err)
```

Comment:
Regarding a previous comment - I missed the fact that the bad connection will still be reported on the span as a warning, so i think between the warning and the attempt field we have all the information we need to monitor the occurrences of ErrBadConn.